### PR TITLE
AP_Compass: add more info as to why Compass Cal failed

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

**Comment by Hwurzburg** [comment]
This is a very big flash hit

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
this is a big code space hit for a one time thing....at a minimum I would code the reason and provide a lookup in the wiki....

**Comment by timtuxworth** [comment]
> this is a big code space hit for a one time thing....at a minimum I would code the reason and provide a lookup in the wiki....

IMO this is a symptom of a bigger problem. In general messages are too cryptic across the board, in some cases it's so bad that to most of us the messages are effectively meaningless. In other cases they are flat out confusing and/or misleading.

There should be a standard way to "code" messages. Is there any chance there is a field available in the message packet that could be used to provide a message number and reason code. 

For this particular PR - I personally think that this information is incredibly helpful and worth every byte. I have been bamboozled by compass calibration failures and I would love to have this level of informative and useful messaging.

**Comment by Hwurzburg** [comment]
@timtuxworth "For this particular PR - I personally think that this information is incredibly helpful and worth every byte."
tell me exactly what you would do if you got a compass cal failure and got this GCS compass cal failure message:??

CompassCal: fit acceptable: true, fix radius: false, calculate orientation: true

followed by six other messages telling if the diag was out of limit or fitness within limits???

how does that help you??what does that tell you to do to fix the problem?

**Comment by timtuxworth** [comment]
> @timtuxworth "For this particular PR - I personally think that this information is incredibly helpful and worth every byte." tell me exactly what you would do if you got a compass cal failure and got this GCS compass cal failure message:??
> 
> CompassCal: fit acceptable: true, fix radius: false, calculate orientation: true
> e
> followed by six other messages telling if the diag was out of limit or fitness within limits???
> 
> how does that help you??what does that tell you to do to fix the problem?

It gives me data I can google. Sure - its not great, but for a better answer, refer to my first comment.

**Comment by Hwurzburg** [comment]
I would like to see what you google and what you get...

**Comment by Hwurzburg** [comment]
I think most if not all of this is sent back to the GCS in a calibration completion status message already and could be expanded into text messages by the GCS and that this is really not required code at all....not an expert, but MAG_CAL_REPORT does exist

**Comment by Hwurzburg** [comment]
Also, the wiki is pretty clear on what to do for a failed compass calibration:
"If calibration fails:

    you will hear an “unhappy” failure tone, the green bars may reset to the left, and the calibration routine may restart (depending upon the ground station). Mission Planner will automatically retry, so continue to rotate the vehicle as instructed above.

   ** if a compass is not calibrating, consider moving to a different area away from magnetic disturbances, and remove electronics from your pockets.** (most the time the reason)

    if, after multiple attempts, the compass has not passed the calibration, Press the “Cancel” button and change the “Fitness” drop-down to a more relaxed setting and try again.

    if compass calibration still fails it may help to raise [COMPASS_OFFS_MAX](https://ardupilot.org/copter/docs/parameters.html#compass-offs-max) from 850 to 2000 or even 3000

    finally, if a single compass is not calibrating and you trust the others, disable it."

**Review by tridge — CHANGES_REQUESTED** [review]
can we just call set_force_log_disarmed() ? and then disable it on end of mag cal
that way we get the full data, and can run magfit

File: libraries/AP_Compass/CompassCalibrator.cpp [hunk 1]
```diff
@@ -189,6 +189,7 @@ void CompassCalibrator::update()
     if (_status == Status::RUNNING_STEP_ONE) {
         if (_fit_step >= 10) {
             if (is_equal(_fitness, _initial_fitness) || isnan(_fitness)) {  // if true, means that fitness is diverging instead of converging
+                GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL, "CompassCal: Calibration routine failed to find solution");
```
[the remark this answers is not in the document]
Comment (timtuxworth) [reply libraries/AP_Compass/CompassCalibrator.cpp:192]: > how about:
> 
> ```
> CompassCal: no solution
> ```

As per my previous comment "As a user, what I'd really like is a clear statement of what is wrong and what I should do about it.", I really don't see how this helps.

**Pull request closed without merging**